hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1597,6 +1597,88 @@ private class Emit(
       case x: NDArrayMap2 =>  emitDeforestedNDArray(x)
       case x: NDArrayReshape => emitDeforestedNDArray(x)
 
+      case NDArrayMatMul(lChild, rChild) =>
+        val lT = emit(lChild)
+        val rT = emit(rChild)
+
+        val lPType = coerce[PNDArray](lChild.pType)
+        val rPType = coerce[PNDArray](rChild.pType)
+
+        val leftND = mb.newField[Long]
+        val rightND = mb.newField[Long]
+
+        val leftShape = lPType.shape.load(region, leftND)
+        val rightShape = rPType.shape.load(region, rightND)
+
+        val lShapeTuple = new CodePTuple(lPType.shape.pType, region, leftShape)
+        val rShapeTuple = new CodePTuple(rPType.shape.pType, region, rightShape)
+
+        val (leftShapeArraySetup, leftShapeArray) = (0 until lPType.nDims).map(i => coerce[Long](lShapeTuple(i))).toArray.cacheEntries(mb, LongInfo)
+        val (rightShapeArraySetup, rightShapeArray) = (0 until rPType.nDims).map(i => coerce[Long](rShapeTuple(i))).toArray.cacheEntries(mb, LongInfo)
+
+        val (unifyShapeSetup, unifiedShapeArray) = NDArrayEmitter.matmulShape(leftShapeArray, rightShapeArray)
+
+        val leftBroadcastMask = if (lPType.nDims > 2) NDArrayEmitter.broadcastMask(leftShapeArray) else Array[Code[Long]]()
+        val rightBroadcastMask = if (rPType.nDims > 2) NDArrayEmitter.broadcastMask(rightShapeArray) else Array[Code[Long]]()
```

Comment:
I think this broadcasting stuff shouldn't get duplicated for each side.

I feel like we should have a class like

```scala
class BroadcastedCodeNDArray(
  val off: Code[Long],
  val t: PNDArray,
  val mb: MethodBuilder
) {
  val (setup, shape) = Array.tabulate(t.nDims)(coerce[Long](lShapeTuple(i))
    .toArray.cacheEntries(mb, LongInfo)
  def loadElementToIRIntermediate(indices: Array[Code[Long]]): Code[_] =
    lPType.loadElementToIRIntermediate(broadcast(indices), off, region, mb)
  private[this] def broadcast(indices: Array[Code[Long]]): Array[Code[Long]] = {
    val stack = (indices zip shape).dropRight(2)
      .map { case (i, s) => (s ceq 0L).mux(0L, s) }
    val matrix = indices.takeRight(2)
    stack ++ matrix
  }
}
```

Reply:
Or we could do the flags thing, but it should get wrapped in a class rather than duplicated everywhere.

Reply:
Is it duplicated everywhere though? It's just these two lines. Other uses of it have different semantics (like not always taking the first 2).

Reply:
I relent, but let's keep this in mind as the code evolves. Something feels wrong about the duplicative nature of the previous 9 lines.